File: NNModule/ModuleManager/Core/ModuleManager.m
//
//  ModuleManager.m
//  ModuleManagment
//
//  Created by NeroXie on 2023/10/27.
//

#import "ModuleManager.h"
#import "Module.h"
#import "ApplicationServiceImpl.h"
#import "URLRouter+NNModule.h"
#import "TabServiceImpl.h"
#import "LaunchTaskServiceImpl.h"
#import "NotificationServiceImpl.h"
#import <objc/runtime.h>
#import <objc/message.h>

@implementation ModuleManager

+ (ModuleManager *)sharedInstance {
    static ModuleManager *_sharedInstance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _sharedInstance = [self new];
    });
    
    return _sharedInstance;
}

- (instancetype)init {
    if (self = [super init]) {
        [Module registerService:@protocol(ModuleApplicationService) usedClass:ApplicationServiceImpl.class];
        [Module registerService:@protocol(ModuleRouteService) usedClass:URLRouter.class];
        [Module registerService:@protocol(ModuleTabService) usedClass:TabServiceImpl.class];
        [Module registerService:@protocol(ModuleLaunchTaskService) usedClass:LaunchTaskServiceImpl.class];
        [Module registerService:@protocol(ModuleNotificationService) usedClass:NotificationServiceImpl.class];
        
        [self _loadAllMethodsFromClass:ModuleRegisterService.class];
    }
    
    return self;
}

- (BOOL)application:(UIApplication *)application willFinishLaunchingWithOptions:(NSDictionary<UIApplicationLaunchOptionsKey,id> *)launchOptions {
    // load application service early
    id<ModuleApplicationService> applicationImpl = Module.applicationService;
    if ([applicationImpl respondsToSelector:@selector(applicationWillAwake)]) {
        [applicationImpl applicationWillAwake];
    }
    
    // wake up all modules.
    [self _loadAllMethodsFromClass:ModuleAwake.class];
    // execute launch task
    id<ModuleLaunchTaskService> launchTaskImpl = Module.launchTaskService;
    if ([launchTaskImpl respondsToSelector:@selector(executeTasks)]) {
        [Module.launchTaskService executeTasks];
    }
    
    if ([applicationImpl respondsToSelector:@selector(application:willFinishLaunchingWithOptions:)]) {
        return [applicationImpl application:application willFinishLaunchingWithOptions:launchOptions];
    }
    
    return YES;
}

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary<UIApplicationLaunchOptionsKey,id> *)launchOptions {
    [self _loadWindowIfNeed];
    
    BOOL result = YES;
    id<ModuleApplicationService> applicationImpl = Module.applicationService;
    if ([applicationImpl respondsToSelector:@selector(application:didFinishLaunchingWithOptions:)]) {
        result = [applicationImpl application:application didFinishLaunchingWithOptions:launchOptions];
    }
    
#if DEBUG
    NSLog(@"application did finish launching");
    [Module serviceInfoPrettyPrinted];
#endif
    return result;
}

- (void)_loadWindowIfNeed {
    if (UIApplication.sharedApplication.keyWindow) return;
    
    id<ModuleApplicationService> applicationImpl = Module.applicationService;
    id<UIApplicationDelegate> applicationDelegate = UIApplication.sharedApplication.delegate;
    UIWindow *window = nil;
    if ([applicationImpl respondsToSelector:@selector(window)]) {
        UIWindow *win = applicationDelegate.window;
        if (win) window = win;
    } else if ([applicationDelegate respondsToSelector:@selector(window)]) {
        UIWindow *win = applicationDelegate.window;
        if (win) window = win;
    } else {
        window = [[UIWindow alloc] initWithFrame:[UIScreen mainScreen].bounds];
        window.backgroundColor = UIColor.whiteColor;
    }
    
    if ([applicationImpl respondsToSelector:@selector(window)]) applicationImpl.window = window;
    if ([applicationDelegate respondsToSelector:@selector(window)]) applicationDelegate.window = window;
    
    [window makeKeyWindow];
}

- (void)_loadAllMethodsFromClass:(Class)aClass {
    Class metaClass = object_getClass(aClass);
    unsigned int count;
    Method *methods = class_copyMethodList(metaClass, &count);
    for (int i = 0; i < count; i++) {
        Method class_Method = methods[i];
        const char *methodName = sel_getName(method_getName(class_Method));
        // Class method of load and initialize invoked by OC runtime
        if (strcmp(methodName, "load") == 0 || strcmp(methodName, "initialize") == 0) continue;
        ((void(*)(id,Method))method_invoke)(metaClass, class_Method);
    }
}

@end

File: NNModule/ModuleManager/Core/ModuleManager.swift
//
//  ModuleManager.swift
//  ModuleManager
//
//  Created by NeroXie on 2019/1/18.
//

import Foundation
import UIKit

@objcMembers public final class ModuleManager: NSObject {
    
    @objc(sharedInstance)
    public static let shared = ModuleManager()
    
    private var registerServices = [Method]()
    
    private var awakes = [Method]()
    
    private override init() {
        super.init()
        
        // register services
        Module.register(service: ModuleRouteService.self, used: URLRouter.self)
        Module.register(service: ModuleTabService.self, used: ModuleTabServiceImpl.self)
        Module.register(service: ModuleLaunchTaskService.self, used: ModuleLaunchTaskServiceImpl.self)
        Module.register(service: ModuleNotificationService.self, used: ModuleNotificationServiceImpl.self)
        Module.register(service: ModuleApplicationService.self, used: ModuleApplicationServiceImpl.self)
        
        // register more services
        loadAllMethods(from: Module.RegisterService.self)
    }
    
    public func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
        loadWindowIfNeed()
        // load application service early
        let applicationImpl = Module.applicationService
        applicationImpl.applicationWillAwake?()
        // wake up all modules to register
        loadAllMethods(from: Module.Awake.self)
        // execute launch task
        Module.launchTaskService.executeTasks()
        
        let result = Module.applicationService.application?(application, didFinishLaunchingWithOptions: launchOptions) ?? true
#if DEBUG
        print("application did finish launching")
        ModuleServiceCenter.shared.serviceInfoPrettyPrinted()
#endif
        return result
    }
    
    private func loadWindowIfNeed() {
        guard UIApplication.shared.keyWindow == nil else { return }
        
        // forcibly load the window
        let sel = NSSelectorFromString("setWindow:")
        var window = UIWindow(frame: UIScreen.main.bounds)
        window.backgroundColor = .white
        if let win = Module.applicationService.window as? UIWindow { window = win }
        Module.applicationService.perform(sel, with: window)
        if let delegate = UIApplication.shared.delegate, delegate.responds(to: sel) {
            delegate.perform(sel, with: window)
        }
        
        window.makeKeyAndVisible()
    }
    
    private func loadAllMethods(from aClass: AnyClass) {
        guard let metaClass: AnyClass = object_getClass(aClass) else { return }
        
        var count: UInt32 = 0
        guard let methodList = class_copyMethodList(metaClass, &count) else { return }
        
        let handle = dlopen(nil, RTLD_LAZY)
        let methodInvoke = dlsym(handle, "method_invoke")
        
        for i in 0..<Int(count) {
            let method = methodList[i]
            unsafeBitCast(methodInvoke, to:(@convention(c)(Any, Method)->Void).self)(metaClass, method)
        }
        
        dlclose(handle)
        free(methodList)
    }
}

